### PR TITLE
[WIP] Test performance of running MIR inliner on inline(always) function calls

### PR DESCRIPTION
**Pull request by wesleywiser**
r? @ghost

### CODE REVIEW

**Comment by wesleywiser** [comment]
A few small wins, lots of significant losses. The main culprit seems to be increased incremental cache load times. 

This is probably proof that addressing #80536 will be critical before we see compile time wins from this optimization.

File: compiler/rustc_mir/src/transform/inline.rs [hunk 1]
```diff
@@ -42,7 +42,7 @@ impl<'tcx> MirPass<'tcx> for Inline {
         // If you change this optimization level, also change the level in
         // `mir_drops_elaborated_and_const_checked` for the call to `mir_inliner_callees`.
         // Otherwise you will get an ICE about stolen MIR.
-        if tcx.sess.opts.debugging_opts.mir_opt_level < 2 {
+        if tcx.sess.opts.debugging_opts.mir_opt_level == 0 {
```
Comment (oli-obk) [inline compiler/rustc_mir/src/transform/inline.rs:45]: could you rerun this experiment, but only enable the optimization in release mode?

**Comment by oli-obk** [comment]
Ok, from the results it looks like we need to turn inlining off during incremental compilation, too.

File: compiler/rustc_mir/src/transform/mod.rs [hunk 2]
```diff
@@ -431,7 +431,7 @@ fn mir_drops_elaborated_and_const_checked<'tcx>(
 
         // Do not compute the mir call graph without said call graph actually being used.
         // Keep this in sync with the mir inliner's optimization level.
-        if tcx.sess.opts.debugging_opts.mir_opt_level >= 2 {
+        if tcx.sess.opts.debugging_opts.mir_opt_level >= 1 {
```
Comment (oli-obk) [inline compiler/rustc_mir/src/transform/mod.rs:434]: To get the performance back in check/debug builds, we'll need to make this conditional on `OptLevel::Aggressive`, too.

**Comment by oli-obk** [comment]
There are a few `-opt` tests that regress, but most of them are either really fast ones, or affected by changes in CGU, let's see how it fares with the two changes mentioned above.

**Comment by tmiasko** [comment]
For comparison, another recent [inlining perf run](https://perf.rust-lang.org/compare.html?start=94736c434ee154b30e2ec22ec112b79e3f6c5884&end=7b783e40d255b2f2881c41fe413bd58bce7e292c) with inlining enabled by default with uniform threshold of 25, and `mir_callgraph_reachable` replaced by `mir_inliner_callees`  (empty iterator is marked as inline so please ignore deeply-nested benchmark).

File: compiler/rustc_mir/src/transform/inline.rs [hunk 3]
```diff
@@ -28,6 +29,21 @@ const UNKNOWN_SIZE_COST: usize = 10;
 
 pub struct Inline;
 
+impl Inline {
+    pub fn should_run_inline_pass<'tcx>(tcx: TyCtxt<'tcx>) -> bool {
```
Comment (oli-obk) [inline compiler/rustc_mir/src/transform/inline.rs:33]: oh heh... that is smart

**Comment by wesleywiser** [comment]
Interesting. I would not have expected this to have an effect on check and debug builds now. Is this just how high the normal variance is?

**Comment by tmiasko** [comment]
From many-assoc-items benchmark:

```
# Baseline  Delta Abs  Shared Object                        Symbol                                                  >
# ........  .........  ...................................  ........................................................>
#
               +5.86%  librustc_driver-50becf5f64b2c1dd.so  [.] rustc_query_system::query::plumbing::try_get_cached
```

Inlining those have been in the past found to be crucial for performance, see #82197.

With inline always using currently no upper cost limit, I expect this is way to aggressive to be practical (hence my earlier attempts with a few statements at most).

**Comment by oli-obk** [comment]
So... we need to go through all `#[inline(always)]` and see whether they are reasonable?

> With inline always using currently no upper cost limit, I expect this is way to aggressive to be practical

Isn't that the point of `always`? Remove the cost limit and just go ahead. The optimization isn't too aggressive in my book, but instead the excessive use of `inline(always)` is.

**Comment by oli-obk** [comment]
or... add more `#[inline(always)]` wherever things now show up more in the benchmarks.

@tmiasko is that snippet you quoted something that can be found on the perf page or do I need to run cachegrind locally?

**Comment by tmiasko** [comment]
> is that snippet you quoted something that can be found on the perf page or do I need to run cachegrind locally?

That is from a script that uses `rustup-toolchain-install-master` and rustc-perf `collector profile_local` to re-run selected benchmarks under your favourite profiler (the output itself is from `perf diff`).

Generally, I would be very conservative about making any MIR optimizations that can have significant performance impact, without having any runtime benchmarks in place.

**Comment by wesleywiser** [comment]
bors seems to have lost this...

@bors try

**Comment by wesleywiser** [comment]
Note, the above run is for a max cost of <= 25.

**Comment by wesleywiser** [comment]
Wow, so somewhere between `25 <= max allowed cost for inline(always) <= 50` we trigger LLVM to go off the deep end when optimizing `deeply-nested` but we also get some really nice wins in real crates:

- `html5ever-opt full` 17.8% win
- `clap-rs-opt full` 13.2% win
- `futures-rs-opt full` 12.9% win
- `cargo-opt full` 8.4% win
- `tokio-webpush-simple-opt full` 8.9% win
- `syn-opt full` 9.6% win
- `style-servo-opt full` 6.4% win
- `web-render-opt full` 7.5% win
- `ripgrep-opt full` 6.4% win

In terms of regressions, other than `deeply-nested-opt`, most of the regressions are < 5% in small benchmarks where LLVM time is usually less than 0.5 seconds. The largest of the regressions are in extremely small benchmarks where the compilation time regresses by tens of milliseconds. 

Interestingly, we're also seeing performance impacts in debug and incremental cases. The incremental release cases could be due to more optimized code coming in from dependencies. I'm less sure about the debug cases. Perhaps std being more optimized is helping in those scenarios?

**Comment by cjgillot** [comment]
On the other hand, the max-rss regression is massive, up to 45% with quite a larger than 25%.
Is there a CGU partitioning effect in the perf improvement?

**Comment by tmiasko** [comment]
| Component       | Before  | After   |
|-----------------|--------:|--------:|
| total           | 375 MiB | 421 MiB |
| rustc-driver.so | 103 MiB |  93 MiB |
| libcore.rlib    |  66 MiB | 100 MiB |
| libstd.rlib     |  15 MiB |  19 MiB |
| libstd.so       |   8 MiB |  10 MiB |

The libcore size increases are primary related to stdarch last time I investiaged the situation, so this should improve soon.

There was a mir-opt-level bump, so this was using the default thresholds. 50 for functions without a hint, 100 for functions with an inline hint, and always inlining with always inline one.

**Comment by cjgillot** [comment]
@wesleywiser are you planning to have this merged? What are the current blockers?

**Comment by mati865** [comment]
AFAIK we were using old LLVM pass manager back when this wad tested. Do you think it'd be worth to retest with the new pass manager?

**Comment by wesleywiser** [comment]
I think it definitely would be worth retesting with NewPM but a naïve rebase ICEs the compiler during bootstrap and I haven't had time to investigate further.

**Pull request closed without merging**